libtransmission/net.cc

```diff
@@ -446,23 +446,28 @@ namespace is_valid_for_peers_helpers
 
 /* isMartianAddr was written by Juliusz Chroboczek,
    and is covered under the same license as third-party/dht/dht.c. */
-[[nodiscard]] auto is_martian_addr(tr_address const& addr)
+[[nodiscard]] auto is_martian_addr(tr_address const& addr, tr_peer_from from)
 {
-    static auto constexpr Zeroes = std::array<unsigned char, 16>{};
+    auto const loopback_allowed = from == TR_PEER_FROM_INCOMING || from == TR_PEER_FROM_LPD || from == TR_PEER_FROM_RESUME;
 
     switch (addr.type)
     {
     case TR_AF_INET:
         {
             auto const* const address = reinterpret_cast<unsigned char const*>(&addr.addr.addr4);
-            return address[0] == 0 || address[0] == 127 || (address[0] & 0xE0) == 0xE0;
+            return address[0] == 0 || // 0.x.x.x
+                (!loopback_allowed && address[0] == 127) || // 127.x.x.x
+                (address[0] & 0xE0) == 0xE0; // multicast address
         }
 
     case TR_AF_INET6:
         {
             auto const* const address = reinterpret_cast<unsigned char const*>(&addr.addr.addr6);
-            return address[0] == 0xFF ||
-                (memcmp(address, std::data(Zeroes), 15) == 0 && (address[15] == 0 || address[15] == 1));
+            return address[0] == 0xFF || // multicast address
+                (std::all_of(address, address + 15, [](char const c) { return c == 0; }) &&
```

Comment:
I imagine that `memcmp` (one comparison) should be faster than `std::all_of` (15 comparisons)?

Reply:
I actually tried the benchmark from lalebarde, including the `std::all_of` solution, and his answer `check_memory_zeroed` is indeed the fastest on my MacBook Pro M3.
Tested with `clang++ -std=c++17 -Os test.cc`.

Reply:
727318c This code is not performance-critical enough to justify the custom function. I'll just keep `memcmp()`.